inference-engine/src/mkldnn_plugin/nodes/mkldnn_topk_node.cpp

```diff
@@ -190,250 +1330,357 @@ std::vector<VectorDims> MKLDNNTopKNode::shapeInfer() const {
-            max_indexes[src_k] = i2;
-            for (int i3 = src_k; i3 > 0; i3--) {
-                if (Compare2<float>()(max_values[i3], max_values[i3 - 1]))
-                    swap_func(i3, i3 - 1);
-                else
-                    break;
+
+        auto layout_dims = dstMemPtr->GetDescWithType<BlockedMemoryDesc>()->getBlockDims();
+        calc_dims_size(layout_dims);
+
+        auto jcp = jit_topk_config_params();
+        jcp.precision = selectedPD->getConfig().inConfs[TOPK_DATA].desc->getPrecision();
+        jcp.data_size = data_size;
+        jcp.blk_size = blk_size;
+        jcp.layout = layout;
+        jcp.top_k = top_k;
+        jcp.axis_dim = axis_dim;
+        jcp.mode_max = mode_max;
+        jcp.sort_index = sort_index;
+        jcp.topk_innermost = topk_innermost;
+        jcp.algorithm = algorithm;
+        jcp.bubble_inplace = bubble_inplace;
+        jcp.sort_stride = static_cast<int>(I);
+        jcp.work_amount = static_cast<int>(I);
+        if (layout == TopKLayoutType::topk_blocked && topk_innermost) {
+            jcp.blk_stride = I * blk_size;
+            if (algorithm == TopKAlgorithm::topk_bubble_sort) {
+                jcp.work_amount = static_cast<int>(axis_dim);
             }
-            s_index += after_num;
         }
-        if (!sort_value) {
-            for (int i2 = 0; i2 < src_k - 1; i2++) {
-                for (int i3 = src_k - 1; i3 > i2; i3--) {
-                    if (std::greater<int>()(max_indexes[i3 - 1], max_indexes[i3])) {
-                        swap_func(i3, i3 - 1);
+
+        if (algorithm == TopKAlgorithm::topk_bitonic_sort) {
+            size_t src_count = srcMemPtr->GetDescWithType<BlockedMemoryDesc>()->getPaddedElementsCount();
+            vec_process_ptr.resize(src_count * data_size);
+            vec_process_idx_ptr.resize(src_count * sizeof(int32_t));
+
+            calc_bitonic_idx(axis_dim, jcp.bitonic_size, jcp.bitonic_idx_cnt, true);
+            if (sort_index) {
+                calc_bitonic_idx(top_k, jcp.bitonic_k_size, jcp.bitonic_k_idx_cnt, false);
+            }
+        }
+
+        if (mayiuse(cpu::x64::avx512_common)) {
+            topk_kernel.reset(new jit_uni_topk_kernel_f32<cpu::x64::avx512_common>(jcp));
+        } else if (mayiuse(cpu::x64::avx2)) {
+            topk_kernel.reset(new jit_uni_topk_kernel_f32<cpu::x64::avx2>(jcp));
+        } else if (mayiuse(cpu::x64::sse41)) {
+            topk_kernel.reset(new jit_uni_topk_kernel_f32<cpu::x64::sse41>(jcp));
+        }
+
+        if (topk_kernel)
+            topk_kernel->create_ker();
+    } else { //reference mode
+        int j;
+        for (j = src_dims.size() - 1; j >= 0; j--) {
+            if (src_dims[j] != 1)
+                break;
+        }
+        if (static_cast<size_t>(j) == axis)
+            is_last_dim = true;
+        dim = static_cast<int>(src_dims[axis]);
+        before_num = count(src_dims, 0, axis);
+    }
+}
+
+void MKLDNNTopKNode::createPrimitive() {
```

Comment:
Is it possible to use MKLDNNNode::createPrimitive() and remove override?

Reply:
Fixed. Thanks Dmitry!